hail/python/hail/expr/aggregators/aggregators.py

```diff
@@ -112,45 +112,92 @@ def _group_by(self, group, agg_expr):
+                rewritten_ir = ir.map_ir(rewrite_base_apply_agg_op)
+
+                seq_op = rewritten_ir.search(lambda x: isinstance(x, SeqOp))
+                assert len(seq_op) == 1
+
+                new_agg_sig = seq_op[0].agg_sig
+                j_agg_sig = Env.hail().expr.Parser.parseAggSignature(str(new_agg_sig))
+
+                return_typ = HailType._from_java(Env.hail().expr.ir.AggOpRegistry.getType(j_agg_sig))
+
+                name = Env.get_uid()
 
-        new_agg_sig = AggSignature(f'Keyed({agg_sig.op})',
-                                    agg_sig.ctor_arg_types,
-                                    agg_sig.initop_arg_types,
-                                    [group.dtype] + agg_sig.seqop_arg_types)
+                keyed_agg_ops.append((name,
+                                              rewritten_ir.__class__(rewritten_ir.a,
+                                                                     rewritten_ir.constructor_args,
+                                                                     rewritten_ir.init_op_args,
+                                                                     new_agg_sig),
+                                              return_typ))
 
-        def rewrite_a(ir):
-            if isinstance(ir, SeqOp):
+                return Apply('get', Ref(name, return_typ), Ref(key, key_typ))
+            elif isinstance(ir, SeqOp):
+                agg_sig = ir.agg_sig
+                new_agg_sig = AggSignature(f'Keyed({agg_sig.op})',
+                                            agg_sig.ctor_arg_types,
+                                            agg_sig.initop_arg_types,
+                                            [group.dtype] + agg_sig.seqop_arg_types)
                 return SeqOp(ir.i, [group._ir] + ir.args, new_agg_sig)
             else:
-                return ir.map_ir(rewrite_a)
+                return ir.map_ir(rewrite_base_apply_agg_op)
 
-        new_a = rewrite_a(a)
+        lifted_agg_op_ir = rewrite_base_apply_agg_op(agg_expr._ir)
 
-        if isinstance(agg_expr._ir, ApplyAggOp):
-            ir = ApplyAggOp(new_a,
-                            ir.constructor_args,
-                            ir.init_op_args,
-                            new_agg_sig)
+        n_agg_ops = len(keyed_agg_ops)
+        assert n_agg_ops != 0
+
+        # Get keys to map over
+        if len(keyed_agg_ops) == 1:
+            first = keyed_agg_ops[0]
+            keys = Apply('keys', Ref(first[0], first[2])) # already sorted
         else:
-            assert isinstance(agg_expr._ir, ApplyScanOp)
-            ir = ApplyScanOp(new_a,
-                             ir.constructor_args,
-                             ir.init_op_args,
-                             new_agg_sig)
+            def key_set(agg_ops, idx):
+                name, ir, typ = agg_ops[idx]
+                if idx != n_agg_ops - 1:
+                    return Apply("union", Apply("keySet", Ref(name, typ)), key_set(agg_ops, idx + 1))
```

Comment:
this is quadratic in number of aggregations

Reply:
I think I need to fix the missing key problem and then every aggregator will have the same set of keys.